kernel/src/dynamic_process_loading.rs

```diff
@@ -0,0 +1,958 @@
+            kernel,
+            chip,
+            flash: Cell::new(flash),
+            app_memory: OptionalCell::empty(),
+            flash_start: Cell::new(0),
+            flash_end: Cell::new(0),
+            fault_policy,
+            new_process_flash: OptionalCell::empty(),
+            driver: driver,
+            buffer: TakeCell::new(buffer),
+            new_app_start_addr: Cell::new(0),
+            new_app_length: Cell::new(0),
+            client: OptionalCell::empty(),
+            current_user: OptionalCell::empty(),
+        }
+    }
+
+    // Needs to be set separately, or breaks grants allocation
+    pub fn flash_and_memory(
+        &self,
+        app_memory: &'static mut [u8],
+        flash_start: usize,
+        flash_end: usize,
+    ) {
+        self.app_memory.set(app_memory);
+        self.flash_start.set(flash_start);
+        self.flash_end.set(flash_end);
+    }
+
+    // Function to find the next available slot in the processes array
+    fn find_open_process_slot(&self) -> Option<usize> {
+        self.procs.map_or(None, |procs| {
+            for (i, p) in procs.iter().enumerate() {
+                if p.is_none() {
+                    return Some(i);
+                }
+            }
+            None
+        })
+    }
+
+    /******************************************* NVM Stuff **********************************************************/
+
+    // Check so see if we are doing something. If not, go ahead and do this
+    // command. If so, this is queued and will be run when the pending
+    // command completes.
+    fn enqueue_command(
+        &self,
+        padding_flag: bool,
+        command: NonvolatileCommand,
+        user_buffer: &'static mut [u8],
+        offset: usize,
+        length: usize,
+    ) -> Result<(), ErrorCode> {
+        //perform this bounds check if it is not a padding header because it comes from the user application
+        if !padding_flag {
+            match command {
+                NonvolatileCommand::UserspaceRead | NonvolatileCommand::UserspaceWrite => {
+                    if offset >= self.new_app_start_addr.get()
+                        || length > self.new_app_length.get()
+                        || offset + length > self.new_app_length.get()
+                    {
+                        return Err(ErrorCode::INVAL);
+                    }
+                }
+            }
+        }
+
+        // Calculate where we want to actually read from in the physical storage.
+        let physical_address = if !padding_flag {
+            offset + self.new_app_start_addr.get() // this offset comes from the app, so it starts from zero.
+                                                   // thererfore we need to add the physical address of the new app
+                                                   // for each write
+        } else {
+            offset // for padding, the kernel passes the address, so no need to add an offset
```

Comment:
This is confusing. Maybe a separate function?

Reply:
I created a separate function for the bounds check and the physical address calculation. It is in a place that I am slightly unhappy with, but it gets the job done.